hydrosdk/predictor.py

```diff
@@ -1,72 +1,165 @@
+            value = getattr(tensor_proto, DTYPE_TO_FIELDNAME[tensor_proto.dtype])
+
+            # If no dims specified in TensorShapeProto, then it is scalar
+            if dims:
+                value = np.reshape(value, dims).tolist()
+                output_tensors_dict[tensor_name] = value
+            else:
+                output_tensors_dict[tensor_name] = value[0]
+
+        return output_tensors_dict
+
+    @staticmethod
+    def predict_resp_to_dict_nparray(response: PredictResponse) -> dict:
+        output_tensors_dict = {}
+        for tensor_name, tensor_proto in response.outputs.items():
+            array_shape = [dim.size for dim in tensor_proto.tensor_shape.dim]
+            np_dtype = proto2np_dtype(tensor_proto.dtype)
+            value = getattr(tensor_proto, DTYPE_TO_FIELDNAME[tensor_proto.dtype])
+            np_array_value = np.array(value, dtype=np_dtype)
+
+            # If no dims specified in TensorShapeProto, then it is scalar
+            if array_shape:
+                output_tensors_dict[tensor_name] = np_array_value.reshape(*array_shape)
+            else:
+                output_tensors_dict[tensor_name] = np.asscalar(np_array_value)
+
+        return output_tensors_dict
+
+    @staticmethod
+    def predict_resp_to_df(response: PredictResponse) -> pd.DataFrame:
+        response_dict: Dict[str, np.array] = PredictServiceClient.predict_resp_to_dict_nparray(response)
+        return pd.DataFrame(response_dict)
+
+
+class Predictable(ABC):
+    @abstractmethod
+    def predictor(self):
+        pass
+
+
+class PredictableApplication(Predictable):
+    """Adds Predictor functionality"""
+
+    def predictor(self, ssl=False, return_type=PredictorDT.DICT_NP_ARRAY) -> PredictServiceClient:
+        """
+
+        :param ssl:
+        :param return_type:
+        :return:
+        """
+        if ssl:
+            self.channel = self.cluster.grpc_secure()
+        else:
+            self.channel = self.cluster.grpc_insecure()
+
+        self.impl = MetricableImplementation(self.channel)
+        self.predictor_return_type = return_type
+
+        return PredictServiceClient(self.impl, self.name, self.model.contract.predict,
+                                    return_type=self.predictor_return_type)
+
+
+class PredictableServable(Predictable):
+    """Adds Predictor functionality"""
+
+    def predictor(self, shadowless=False, ssl=False, return_type=PredictorDT.DICT_NP_ARRAY) -> PredictServiceClient:
+        """
+
+        :param shadowless:
+        :param ssl:
+        :param return_type:
+        :return:
+        """
+        if ssl:
+            self.channel = self.cluster.grpc_secure()
+        else:
+            self.channel = self.cluster.grpc_insecure()
+        if shadowless:
+            self.impl = UnmetricableImplementation(self.channel)
+        else:
+            self.impl = MetricableImplementation(self.channel)
+        self.predictor_return_type = return_type
```

Comment:
To be frank, my suggestion to use the word `shadowless` is bad. It's hard to reason about the code, when there is double negation in monitoring logic.

Perhaps, we should change the parameter to `monitorable`?
```suggestion
    def predictor(self, monitorable=True, ssl=False, return_type=PredictorDT.DICT_NP_ARRAY) -> PredictServiceClient:
        """

        :param monitorable:
        :param ssl:
        :param return_type:
        :return:
        """
        if ssl:
            self.channel = self.cluster.grpc_secure()
        else:
            self.channel = self.cluster.grpc_insecure()
        if monitorable:
            self.impl = MetricableImplementation(self.channel)
        else:
            self.impl = UnmetricableImplementation(self.channel)
        self.predictor_return_type = return_type
```

Reply:
added